### PR TITLE
Pyn project relate sale purchase

### PR DESCRIPTION
**Pull request by phongyanon**
please review my code.

### CODE REVIEW

File: specific/cmo_project/models/project.py [hunk 1]
```diff
@@ -293,15 +315,116 @@ def _get_state_before_inactive(self):
...
+    @api.multi
+    @api.depends(
+        'actual_price',
+        'estimate_cost',
+        'quote_related_ids',
+        'quote_related_ids.order_line',
+    )
+    def _compute_price_and_cost(self):
+        for project in self:
+            actual_price = 0
```
Comment (kittiu) [inline specific/cmo_project/models/project.py]: Please declare as float, 0.0
This is a good practice.

File: specific/cmo_project/models/project.py [hunk 2]
```diff
@@ -293,15 +315,116 @@ def _get_state_before_inactive(self):
...
+            )
+            project.actual_po = sum(purchase_orders.mapped('amount_untaxed'))
+
+    @api.multi
+    def name_get(self):
+        res = []
+        for project in self:
+            name = project.name or '/'
+            if name and project.project_number:
+                name = name + ' ('+project.project_number+')'
```
Comment (kittiu) [inline specific/cmo_project/models/project.py]: Always use string injection, not "+" to concatenate string, i..e,
`'%s (%s)' % (name, project.project_number)`

File: specific/cmo_project/models/project.py [hunk 3]
```diff
@@ -94,6 +100,7 @@ class ProjectProject(models.Model):
     expense = fields.Float(
         string='Expense',
         states={'close': [('readonly', True)]},
+        compute='_compute_expense',
     )
     brief_date = fields.Date(
```
Comment (kittiu) [inline specific/cmo_project/models/project.py]: name it "date_brief" is better as it is our standard.
But non-blocking.

File: specific/cmo_project/models/project.py [hunk 4]
```diff
@@ -293,15 +315,116 @@ def _get_state_before_inactive(self):
     @api.multi
     def _set_project_analytic_account(self):
         for project in self:
-            parent_project = self.env['project.project'].browse(project.project_parent_id.id)
+            parent_project = self.env['project.project'].browse(
+                project.project_parent_id.id)
             project.parent_id = parent_project.analytic_account_id.id
```
Comment (kittiu) [inline specific/cmo_project/models/project.py]: I am not sure about this.
Is "parent_id" and analytic_account? If so, it is ok.
Also,
Normally you don't need .id to assign m2o to m2o. , i..e, `project.location_id = purchase.order_line.locatoin_id`

File: specific/cmo_project/models/project.py [hunk 5]
```diff
@@ -293,15 +315,116 @@ def _get_state_before_inactive(self):
...
 
     @api.multi
     @api.constrains('brief_date', 'date')
     def _check_brief_dates(self):
         self.ensure_one()
         if self.brief_date > self.date:
-            return ValidationError("project brief-date must be lower than project end-date.")
+            return ValidationError("project brief-date must be lower than \
```
Comment (kittiu) [inline specific/cmo_project/models/project.py]: For long message, python can write like this.
```
return ValidationError("project brief-date must be lower than"
                       "project end-date.")
```
Not hat we can new line without \ or +

File: specific/cmo_project/models/project.py [hunk 6]
```diff
@@ -293,15 +315,116 @@ def _get_state_before_inactive(self):
...
+                project end-date.")
+
+    @api.multi
+    def quotation_relate_project_tree_view(self):
+        self.ensure_one()
+        domain = [
+            '&',
+            ('project_related_id', 'like', self.id),
```
Comment (kittiu) [inline specific/cmo_project/models/project.py]: 'like', self.id
not
'=', self.id   ?
I think we can not use like with and id

File: specific/cmo_project/models/project.py [hunk 7]
```diff
@@ -293,15 +315,116 @@ def _get_state_before_inactive(self):
...
     @api.constrains('brief_date', 'date')
     def _check_brief_dates(self):
         self.ensure_one()
         if self.brief_date > self.date:
-            return ValidationError("project brief-date must be lower than project end-date.")
+            return ValidationError("project brief-date must be lower than \
+                project end-date.")
+
+    @api.multi
+    def quotation_relate_project_tree_view(self):
```
Comment (kittiu) [inline specific/cmo_project/models/project.py]: For a button click that redirect to another action window. If the action already available, you can use (sample)
```
    @api.multi
    def open_some_action_window(self):
        self.ensure_one()
        action = self.env.ref('account_budget_activity.'
                              'act_account_budget_view')
        result = action.read()[0]
        dom = ['|', ('ref_budget_id', '=', self.id), ('id', '=', self.id)]
        result.update({'domain': dom})
        return result
```

File: specific/cmo_project/models/project.py [hunk 8]
```diff
@@ -293,15 +315,116 @@ def _get_state_before_inactive(self):
...
+            'type': 'ir.actions.act_window',
+            'view_type': 'form',
+            'view_mode': 'tree,form',
+            'domain': domain,
+            'context': "{'active': True, 'order_type': 'quotation'}",
+        }
+
+    @api.multi
+    def purchase_relate_project_tree_view(self):
+        self.ensure_one()
```
Comment (kittiu) [inline specific/cmo_project/models/project.py]: Use existing action if exists.

**Comment by kittiu** [comment]
@phongyanon please fix accordingly and PR again, i will merge. Thanks,

File: specific/cmo_project/models/project.py (hunk 1 above)
Comment (kittiu) [reply specific/cmo_project/models/project.py]: This one is not fixed?

File: specific/cmo_project/models/project.py (hunk 3 above)
Comment (kittiu) [reply specific/cmo_project/models/project.py]: This on is not fixed?

File: specific/cmo_project/models/project.py (hunk 4 above)
Comment (kittiu) [reply specific/cmo_project/models/project.py]: This one is not fixed? If it won't fix, please also comment so i will know.

File: specific/cmo_project/models/project.py (hunk 6 above)
Comment (kittiu) [reply specific/cmo_project/models/project.py]: This one is not fixed?

File: specific/cmo_project/models/project.py (hunk 8 above)
Comment (kittiu) [reply specific/cmo_project/models/project.py]: This one is not fixed?

File: specific/cmo_project/models/project.py (hunk 7 above)
Comment (kittiu) [reply specific/cmo_project/models/project.py]: This one?

**Comment by phongyanon** [comment]
Ready to merge

**Pull request merged**